src/particles/wakefields/ChargeBinning.cpp

```diff
@@ -0,0 +1,191 @@
+
+                        // Divide charge by bin size to get binned charge density
+                        amrex::Real add_value = ablastr::constant::SI::q_e / bin_size;
+
+                        // Calculate the charge contribution of the macro particle
+                        if (!is_unity_particle_weight)
+                        {
+                            add_value *= w;
+                        }
+
+                        // Add to histogram bin
+                        amrex::HostDevice::Atomic::Add(&dptr_data[bin], add_value);
+                    });
+                }
+            }
+        }
+    }
+
+    void DerivativeCharge1D (
+        amrex::Gpu::DeviceVector<amrex::Real> const & charge_distribution,
+        amrex::Gpu::DeviceVector<amrex::Real> & slopes,
+        amrex::Real bin_size,
+        bool GetNumberDensity
+    )
+    {
+        int const num_bins = charge_distribution.size();
+        amrex::Real const * const dptr_charge_distribution = charge_distribution.data();
+        amrex::Real * const dptr_slopes = slopes.data();
+
+        amrex::ParallelFor(num_bins - 1, [=] AMREX_GPU_DEVICE(int i)
+        {
+            // Compute the charge density derivative
+            amrex::Real const charge_derivative = (dptr_charge_distribution[i + 1] - dptr_charge_distribution[i]) / bin_size;
+
+            // If GetNumberDensity = True, convert charge density derivative to number density derivative for CSR convolution
+            if (GetNumberDensity)
+            {
+                dptr_slopes[i] = charge_derivative / ablastr::constant::SI::q_e;
+            }
+            else
+            {
+                dptr_slopes[i] = charge_derivative;
+            }
+        });
+    }
+
+    void MeanTransversePosition (
+        impactx::ImpactXParticleContainer& myspc,
+        amrex::Gpu::DeviceVector<amrex::Real> & mean_x,
+        amrex::Gpu::DeviceVector<amrex::Real> & mean_y,
+        amrex::Real bin_min,
+        amrex::Real bin_size,
+        bool is_unity_particle_weight
+    )
+    {
+        using namespace amrex::literals;
+
+
+
+        int const num_bins = mean_x.size();
+        amrex::Real* dptr_mean_x = mean_x.data();
+        amrex::Real* dptr_mean_y = mean_y.data();
+
+        // Declare arrays for sums of positions and weights
+        auto sum_x = amrex::Gpu::DeviceVector<amrex::Real>(num_bins, 0.0_rt);
+        auto sum_y = amrex::Gpu::DeviceVector<amrex::Real>(num_bins, 0.0_rt);
+        auto sum_w = amrex::Gpu::DeviceVector<amrex::Real>(num_bins, 0.0_rt);
+
+        amrex::Real* const sum_w_ptr = sum_w.data();
+        amrex::Real* const sum_x_ptr = sum_x.data();
+        amrex::Real* const sum_y_ptr = sum_y.data();
+
+        int const nlevs = myspc.finestLevel();
+        for (int lev = 0; lev <= nlevs; ++lev)
+        {
+            #ifdef AMREX_USE_OMP
+            #pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
+            #endif
```

Comment:
This should be a race condition because `sum_*` arrays are not defined thread local.
```suggestion
```